site/docs/expressions/scalar_functions.md

```diff
@@ -1,52 +1,118 @@
 
-| Property        | Description                                                  |
-| --------------- | ------------------------------------------------------------ |
-| Organization Id | An unsigned 32 bit integer mapped to a list of known organizations listed in the [Substrait repository](https://github.com/substrait-io/substrait/blob/main/extensions/organizations.yaml). |
-| Function Id     | An unsigned 32 bit integer mapped to a list of known functions for the specific organization. For the Substrait organization, the function is listed in [here](https://github.com/substrait-io/substrait/blob/main/extensions/scalar_functions.yaml). |
+### Parameterized Types
 
-An organization is responsible for managing their own lists of functions. A function id should be used forever. If a function is deleted for any reason, the id should not be reused. Ideally each function signature will have one or more 
+Types are parametered by two types of values: by inner types (e.g. List&lt;K&gt;) and numeric values (e.g. VARCHAR(P,S)). Parameter names are simple strings (frequently a single character). There are two types of parameters: integer parameters and type parameters. 
 
-## Scalar Function Bindings
+When the same parameter name is used multiple times in a function definition, the function can only bind if the exact same value is used for all parameters of that name. For example, if one had a function with a signature of `fn(VARCHAR(N), VARCHAR(N)), the function would be only be usable if both VARCHAR types had the same length value N. This necessitates that all instances of the same parameter name must be of the same parameter type (all instances are a type parameter or always intances are aninteger parameter).
 
-For scalar functions (one value produced for each record), a function binding declares the function identifier to be used and the input arguments.
+#### Integer Parameter Bounds
 
+A type which declares one or more integer parameters can also declare a range of legal values for a function. For example, one could declare a function `fn(VARCHAR(N)) where N IN [0..20]` to describe a function which only takes in a varchar types that are less than or equal to 20 characters in length.
+
+#### Type Parameter Bounds
+
+A parameterized type can either be unbounded (any type is allowed) or be bounded. When bounded, the function can only be bound to arguments of the types bound to. For example, one could declare a function `fn(Map<K,V> map) where K IN [STRING, VARCHAR(N), FIXEDCHAR(N)]` which would allow the function to accept a map argument as long as the key type is one of STRING, VARCHAR or FIXEDCHAR. Note, when duplicate parameter names are used in disjunct bounds options, the names will use the same bounds. For example, in the above map function, if one were to declare a bound such as where N IN [0..20], this limit would apply to either N parameter.
```

Comment:
If I have `fn(K, K) where K IN [STRING, VARCHAR(N), FIXEDCHAR(N)]` is `fn(STRING, VARCHAR(N))` a valid combination?

Reply:
No. My goal with the wording previously was that K must the exact same value in all locations. If you want to support the pattern of mixed arguments, you'd need to use separate parameter names.